UM/Settings/ContainerQuery.py

```diff
@@ -0,0 +1,133 @@
+        self._registry = registry
+
+        self._container_type = container_type
+        self._ignore_case = ignore_case
+        self._kwargs = kwargs
+
+        self._result = None
+
+    ##  Retrieve the result of this query.
+    #
+    #   \return A list of containers matching this query, or None if the query was not executed.
+    def getResult(self):
+        return self._result
+
+    ##  Check to see if this is a very simple query that looks up a single container by ID.
+    #
+    #   \return True if this query is case sensitive, has only 1 thing to search for and that thing is "id".
+    def isIdOnly(self):
+        return not self._ignore_case and len(self._kwargs) == 1 and "id" in self._kwargs
+
+    ##  Execute the actual query.
+    #
+    #   This will search the containers of the ContainerRegistry based on the arguments provided to this
+    #   class' constructor. After it is done, the result can be retrieved with getResult().
+    def execute(self):
+        containers = []
+        for container in filter(lambda c: not self._container_type or isinstance(c, self._container_type), self._registry._containers):
+            matches_container = True
+            for key, value in self._kwargs.items():
+                if isinstance(value, str):
+                    if "*" in value:
+                        matches_container = self._matchRegExp(container, key, value)
+                    else:
+                        matches_container = self._matchString(container, key, value)
+                else:
+                    matches_container = self._matchType(container, key, value)
+
+                if not matches_container: # For the moment, a query needs to match all specified criteria
+                    break
+
+            if matches_container:
+                containers.append(container)
+
+        self._result = containers
+
+    def __hash__(self):
+        return hash(self.__key())
+
+    def __eq__(self, other):
+        return self.__key() == other.__key()
+
+    # protected:
+
+    # Check to see if a container matches with a regular expression
+    def _matchRegExp(self, container, property_name, value):
+        value = re.escape(value) #Escape for regex patterns.
+        value = "^" + value.replace("\\*", ".*") + "$" #Instead of (now escaped) asterisks, match on any string. Also add anchors for a complete match.
+        if self._ignore_case:
+            value_pattern = re.compile(value, re.IGNORECASE)
+        else:
+            value_pattern = re.compile(value)
+
+        if property_name == "id":
+            return value_pattern.match(container.getId())
+        elif property_name == "name":
+            return value_pattern.match(container.getName())
+        elif property_name == "definition":
+            try:
+                return value_pattern.match(container.getDefinition().getId())
+            except AttributeError:  # Only instanceContainers have a get definition. We can ignore all others.
```

Comment:
It was code copied from the original implementation in ContainerRegistry. But I have now replaced those statements with an isinstance check.